### ISSUE TITLE
Versioning (Constraint) Support

### ISSUE THREAD

**Issue by ngRevan** [rbeauchamp/Swashbuckle.OData#124]
In my OData Service I use the [Api-Versioning library from Microsoft](https://github.com/Microsoft/aspnet-api-versioning).

Issue in versioning repo: Microsoft/aspnet-api-versioning#60

I register the OData Service with an Api Version constraint: 
`config.MapVersionedODataRoutes("odata", "api/v{apiVersion}", modelBuilder.GetEdmModels());`

The library creates the OData routes according to the model version supplied and creates a custom OData Route Constraint which validates the given version with the version of the model.

The problem is that the ODataSwaggerProvider fails to generate a valid route and throws a System.ArgumentNullException which results in a 500 Status Error. Stack Trace below.

```
500 : {
    "Message":"An error has occurred.",
    "ExceptionMessage":"Value cannot be null. Parameter name: routeData",
    "ExceptionType":"System.ArgumentNullException",
    "StackTrace":" 
        at System.Net.Http.HttpRequestMessageExtensions.SetRouteData(HttpRequestMessage request, IHttpRouteData routeData)
        at Swashbuckle.OData.Descriptions.SwaggerRouteStrategy.CreateHttpRequestMessage(HttpMethod httpMethod, Operation potentialOperation, SwaggerRoute potentialSwaggerRoute, HttpConfiguration httpConfig)
        at Swashbuckle.OData.Descriptions.SwaggerRouteStrategy.GetActionDescriptors(HttpMethod httpMethod, Operation potentialOperation, SwaggerRoute potentialSwaggerRoute, HttpConfiguration httpConfig)
        at Swashbuckle.OData.Descriptions.SwaggerRouteStrategy.GetActionDescriptors(SwaggerRoute potentialSwaggerRoute, HttpConfiguration httpConfig)
        at Swashbuckle.OData.Descriptions.SwaggerRouteStrategy.<>c__DisplayClass3_0.<Generate>b__1(SwaggerRoute potentialSwaggerRoute)
        at System.Linq.Enumerable.<SelectManyIterator>d__14`2.MoveNext()
        at System.Linq.Enumerable.<SelectManyIterator>d__14`2.MoveNext()
        at System.Linq.Enumerable.<DistinctIterator>d__81`1.MoveNext()
        at System.Linq.Enumerable.<SelectManyIterator>d__14`2.MoveNext()
        at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
        at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
        at Swashbuckle.OData.CollectionExtentions.ToCollection[T](IEnumerable`1 source)
        at Swashbuckle.OData.Descriptions.ODataApiExplorer.GetApiDescriptions()
        at System.Lazy`1.CreateValue()
        at System.Lazy`1.LazyInitValue()
        at System.Lazy`1.get_Value()
        at Swashbuckle.OData.Descriptions.ODataApiExplorer.get_ApiDescriptions()
        at Swashbuckle.OData.ODataSwaggerProvider.GetApiDescriptionsFor(String apiVersion)
        at Swashbuckle.OData.ODataSwaggerProvider.GetSwagger(String rootUrl, String apiVersion)
        at Swashbuckle.Application.SwaggerDocsHandler.SendAsync(HttpRequestMessage request,     CancellationToken cancellationToken)
        at System.Net.Http.HttpMessageInvoker.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
        at System.Web.Http.Dispatcher.HttpRoutingDispatcher.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
        at System.Net.Http.DelegatingHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
        at System.Web.Http.HttpServer.<>n__FabricatedMethod9(HttpRequestMessage , CancellationToken )
        at System.Web.Http.HttpServer.<SendAsync>d__0.MoveNext()"
}     http://localhost:8080/swagger/docs/v1
```

In the ODataPathHandler the service root should contain the api version (which is known to swagger) and the ODataPath should not contain the api version as a query parameter or should at least supply a valid value for the api version.

```csharp
    public class CustomODataPathHandler : DefaultODataPathHandler
    {
        public override ODataPath Parse(IEdmModel model, string serviceRoot, string odataPath)
        {
            // serviceRoot = "http://any/" should be "http://any/v1"
            // odataPath = "ExampleResource?apiVersion='SampleString'" should be "ExampleResource"
        }
    }
```

**Comment by alexandery** [comment]
+1. The issues is still present.

**Comment by tsharp** [comment]
+1

**Comment by extinctPencil** [comment]
I faced this also . 
I tried to resolve the code myself ... The issue is that a Version path parameter resolver  would need a reference to Microsoft.ASPNet.OData.Versioning
I'm sure there are better ways of doing this via dependency Injection  but for what its worth
here is what I did 
**Modified** 
Swashbuckle.OData.Descriptions.OperationExtensions ==> GenerateSampleODataUri


            if (parameters != null && parameters.Any())
            {
                // if we have been supplied a versionPathResolver lets Use it.
               ODataSwaggerProvider.VersionPathParameterResolver?.Invoke(parameters, route);


**Modified**
Swashbuckle.OData.ODataSwaggerProvider  (declare and use a delegate)

namespace Swashbuckle.OData
{

    public delegate void VersionPathParameterResolver(IDictionary<string, string> parameters, ODataRoute route);

    public class ODataSwaggerProvider : ISwaggerProvider
    {
        private readonly ISwaggerProvider _defaultProvider;
        private readonly ODataSwaggerDocsConfig _config;

        private static ConcurrentDictionary<string, Lazy<SwaggerDocument>> _cache =
            new ConcurrentDictionary<string, Lazy<SwaggerDocument>>();


        public static VersionPathParameterResolver VersionPathParameterResolver;


Then in **My** Service project Which does have a reference to  Microsoft.ASPNet.OData.Versioning provide a Delegate that resolves the version

`ODataSwaggerProvider.VersionPathParameterResolver = UpdateVersionPathParameter;`


```
  /// <summary>
        /// Introduced Method to resolve Version paramters 
        /// </summary>
        /// <param name="parameters"></param>
        /// <param name="route"></param>
        public static void UpdateVersionPathParameter(IDictionary<string, string> parameters, ODataRoute route)
        {
            string apiVersionKey = "apiVersion";
            string oDataConstraintKey = "ODataConstraint";

            if (parameters != null && parameters.Any())
            {
                if (parameters.ContainsKey(apiVersionKey))
                {

                    if (route != null && route.Constraints.Any())
                    {
                        if (route.Constraints.ContainsKey(apiVersionKey))
                        {
                            if (route.Constraints.ContainsKey(oDataConstraintKey))
                            {
                                var versionConstraint = route.Constraints[oDataConstraintKey] as VersionedODataPathRouteConstraint;
                                if (versionConstraint != null)
                                {
                                    var majVer = versionConstraint.ApiVersion.MajorVersion;
                                    if (majVer.HasValue)
                                    {
                                        parameters[apiVersionKey] = majVer.Value.ToString();
                                    }
                                }

                            }
                        }
                    }

                }
            }
        }
```

**Comment by vanikulkarniAtFIS** [comment]
+1. The issue is still present.

**Comment by vanikulkarniAtFIS** [comment]
We badly need the issue fixed in the package. Is anyone doing active development on this bug?

**Comment by extinctPencil** [comment]
Hi I dont really have the capacity to move this forward 
I have Created a Fork with these changes along with the necessary changes for the latest versions of OData. I am using this fork in my projects without issue so far.  I do feel however that my solutions to the problems could be refactored and a more elegant / extensible  way (DI rather than a delegate for example). I solved the problems I faced directly , treating the symtoms rather than the cause. There could be more problems and there is the possibility that the solutions offered impact some other use case.
Hopefully someone with more time or experience in this area can adopt them
Best regards
